Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+              case 2 where part.first == "." && part.last == ".":
+                // If there's a previous part, drop it; otherwise, do nothing.
+                if let prev = parts.last {
+                    parts.removeLast()
+                    capacity -= prev.count
+                }
+              default:
+                // Any other component gets appended.
+                parts.append(part)
+                capacity += part.count
+            }
+        }
+        capacity += max(parts.count, 1)
+        
+        // Create an output buffer using the capacity we've calculated.
+        // FIXME: Determine whether this is the most efficient way to do it.
+        var result = ""
+        result.reserveCapacity(capacity)
+        
+        // Put the normalized parts back together again.
+        var iter = parts.makeIterator()
+        result.append(PathSeparatorCharacter)
+        if let first = iter.next() {
+            result.append(contentsOf: first)
+            while let next = iter.next() {
+                result.append(PathSeparatorCharacter)
+                result.append(contentsOf: next)
+            }
+        }
+        
+        // Sanity-check the result (including the capacity we reserved).
+        assert(!result.isEmpty, "unexpected empty string")
+        assert(result.characters.count == capacity, "count: " +
+            "\(result.characters.count), cap: \(capacity)")
+        
+        // Use the result as our stored string.
+        _impl = PathImpl(string: String(result))
+    }
+    
+    /// Directory component.  An absolute path always has a non-empty directory
+    /// component (the directory component of the root path is the root itself).
+    public var dirname: String {
+        return _impl.dirname!
+    }
+    
+    /// Last path component (including the suffix, if any).  it is never empty.
+    public var basename: String {
+        return _impl.basename
+    }
+    
+    /// Suffix (including leading `.` character) if any.  Note that a basename
+    /// that starts with a `.` character is not considered a suffix.
+    public var suffix: String? {
+        return _impl.suffix
+    }
+    
+    /// Absolute path of parent directory.  This always returns a path, because
+    /// every directory has a parent (the parent directory of the root directory
+    /// is considered to be the root directory itself).
+    public var parentDirectory: AbsolutePath {
+        return isRoot ? self : AbsolutePath(_impl.dirname!)
+    }
+    
+    /// True if the path is the root directory.
+    public var isRoot: Bool {
+        let chars = _impl.string.characters
+        return chars.count == 1 && chars.first == PathSeparatorCharacter
+    }
+    
+    /// Returns the absolute path with the relative path applied.
+    public func join(_ subpath: RelativePath) -> AbsolutePath {
+        let slash = String(PathSeparatorCharacter)
+        return AbsolutePath(asString + slash + subpath.asString)
```

Comment:
What I was specifically referring to was a fast path over `asString + slash + subpath.asString` to not re-normalize the starting path... the comment below seemed to be about something else.